teps/0117-manage-pipelinerun-concurrency.md

```diff
@@ -0,0 +1,368 @@
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+  - [Reusability](#reusability)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [User Experience](#user-experience)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+Manage pipeline run concurrency via configurable strategies.
+
+## Motivation
+There are situations where pipelines can overlap, and they should not. For instance, a pr is created and that triggers ci to so a PipelineRun is created, a developer notices a small typo, pushed again, and a new PipelineRun is triggered, the first PipelineRun is no longer relevant, but potentially still running.  
+Another example is managing available shared resources, lets say tekton triggers an external device and there are a set available amount, or simply resource constrained clusters. 
+
+
+2 proposed cancel strategies. 
+- ZeroOverlap:
+  - Pipelineruns with the same key do not overlap, so if one is running and a new one is scheduled, the former is canceled.
+  - Configure cancel status ( with finally, stop, cancel )
+- Queues:
+  - Queues have a max run ( how many can run at the same time )
+  - fifo
+  - Queue max can be over-ridden, but is always the last pipelinerun created.     
+
+
+Strategies can be mixed together BUT their concurrency keys cannot be the same ( cancel eachother out ).
+
+Mixing strategies would allow for better resource management for teams. For instance, 3 teams are working on a mobile application with only 5 test devices so our max queue size is 5. If team 1 works in really fast bursts, 
+they might clog the queue with small pr changes. Leaving team 2 & 3 only able to get a build in when lucky. With zeroOverlap, and a queue, a single pr with many quick changes will only ever occupy one slot in the queue.    
+
+### Requirements
+- PipelineRuns are canceled based on the strategy defined.
+- Simple interface set at the PipelineRun manifest.
+- New PipelineRun should wait for the canceled one to finish when applicable.
+- All PipelineRuns follow fifo patterns. 
+
+## Proposal
+Adding a key on the PipelineRun manifest seems like the most appropriate, and simple solution.  
+  
+Short example:  
+```yaml
+concurrency:
```

Comment:
potential alternative syntax:

```
concurrency:
   key: $(params.foo)
   strategy: cancel # options here are "queue" and "cancel"
   max: 1  # we can choose a default value for max depending on the concurrency strategy
```

Reply:
This alternative syntax does not give the ability to use both a queue and zeroOverlap at the same time right? or am i misunderstanding part of it?

Reply:
It doesn't. Imagine there are 5 PipelineRuns running, with max = 5. My understanding would be that if a new PipelineRun was created and the strategy was "queue", it would wait until one finished before starting. If the strategy was "cancel", it would cancel the earliest created PipelineRun and then start. ZeroOverlap is "cancel" with max = 1.

I'm not understanding why you might want to use both at the same time; they seem incompatible with each other. Can you describe what you had in mind?